scrapy_webarchive/utils.py

```diff
@@ -78,3 +77,23 @@ def add_ftp_credentials(wacz_uri: str, settings: Settings) -> str:
         return urlunparse(updated_uri)
     
     return wacz_uri
+
+
+def hash_stream(hash_type: str, stream: IO):
+    """Hashes the stream with given hash_type hasher"""
+
+    try:
+        hasher = hashlib.new(hash_type)
+    except (ImportError, ValueError):
+        return 0, ""
```

Comment:
When would `ImportError` happen? If the hash type is not loaded? Some error message / at least a warning would be appropriate.

Reply:
> "... so we just want to know if we can use this algorithm (and it doesn't matter if a different platform would not support this)."

But it **does** matter if a different platform supports this or not. We cannot assume the platform these spiders are deployed to. People could introduce OS-specific options that would not be supported on a different platform. Using `hashlib.algorithms_guaranteed` would be safe to use for maximum compatibility. I would suggest leaving this as is.

Reply:
Ah, that is a good point: when we generate an archive, we would like the hash to be readable by the user of the archive.
But then we would need to take into account _all_ software that may handle this, not just Python.

Does datapackage / WACZ mention anything about hasing algorithms?

Reply:
In their examples it is mainly `sha256` that is used, but it does not mention anything about it in the specification. `sha256` is also what is hardcoded as a variable in the `WaczFileCreator` class where the `hash_stream` function is used, without any way to override it (except for subclassing the `WaczFileCreator` class. But if we want to make this adjustable through a setting, it would definitely be good to have a solid implementation of the `hash_stream` function.

Reply:
Decided to raise an exception instead of logging and returning early. Returning size 0 and an empty hash would be misleading because a user would have to lookup the logs in case something went wrong. Seems to be more appropriate to just raise an exception here.